File: ios-native/NeobankNative/Sources/Features/Notifications/NotificationRepository.swift
import Foundation

struct NotificationRepository: Sendable {
    let client: APIClient

    private struct MarkAllReadResponse: Decodable, Sendable {
        let markedCount: Int
    }

    func list(cursor: String? = nil, limit: Int = 20) async throws -> NotificationPage {
        try await client.send("/v1/notifications", query: ["limit": String(limit), "cursor": cursor])
    }

    func markRead(_ id: String) async throws -> AppNotification {
        try await client.send("/v1/notifications/\(id)/read", method: .post)
    }

    @discardableResult
    func markAllRead() async throws -> Int {
        let response: MarkAllReadResponse = try await client.send("/v1/notifications/read-all", method: .post)
        return response.markedCount
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/notifications/NotificationRepository.kt
package dev.horobets.neobanknative.features.notifications

import dev.horobets.neobanknative.core.networking.ApiClient
import dev.horobets.neobanknative.core.networking.HttpMethod
import kotlinx.serialization.Serializable

class NotificationRepository(private val client: ApiClient) {
    @Serializable
    private data class MarkAllReadResponse(val markedCount: Int)

    suspend fun list(cursor: String? = null, limit: Int = 20): NotificationPage = client.send(
        "/v1/notifications",
        query = mapOf("limit" to limit.toString(), "cursor" to cursor),
    )

    suspend fun markRead(id: String): AppNotification = client.send("/v1/notifications/$id/read", method = HttpMethod.POST)

    suspend fun markAllRead(): Int =
        client.send<MarkAllReadResponse>("/v1/notifications/read-all", method = HttpMethod.POST).markedCount
}
